timer/node_timer.go

```diff
@@ -0,0 +1,143 @@
+	parentComponent *component
+
+	// parent is the parent timer. It is used to propagate the stop of the timer to the parent timer and get the remaining time in case the budget has to be inherited.
+	parent NodeTimer
+}
+
+var _ NodeTimer = (*nodeTimer)(nil)
+
+// NewTreeTimer creates a new Timer with the given options. You have to specify either the option WithBudget or WithUnlimitedBudget and at least one component using the WithComponents option.
+func NewTreeTimer(options ...Option) (NodeTimer, error) {
+	config := newConfig(options...)
+	if config.Budget == InheritedBudget {
+		return nil, errors.New("root timer cannot inherit parent budget, please provide a budget using timer.WithBudget() or timer.WithUnlimitedBudget()")
+	}
+
+	if len(config.Components) == 0 {
+		return nil, errors.New("NewTreeTimer: tree timer must have at least one component, otherwise use NewTimer()")
+	}
+
+	return &nodeTimer{
+		baseTimer: baseTimer{
+			config:    config,
+			timeCache: newTimeCache(),
+		},
+		components: newComponents(config.Components),
+	}, nil
+}
+
+func (timer *nodeTimer) NewNode(name string, options ...Option) (NodeTimer, error) {
+	config := newConfig(options...)
+	if len(config.Components) == 0 {
+		return nil, errors.New("NewNode: node timer must have at least one component, otherwise use NewLeaf()")
+	}
+
+	component, ok := timer.components.lookup[name]
+	if !ok {
+		return nil, fmt.Errorf("NewNode: component %s not found", name)
+	}
+
+	return &nodeTimer{
+		baseTimer: baseTimer{
+			config:    config,
+			timeCache: timer.timeCache,
+		},
+		components:      newComponents(config.Components),
+		parentComponent: component,
+		parent:          timer,
+	}, nil
+}
+
+func (timer *nodeTimer) NewLeaf(name string, options ...Option) (Timer, error) {
+	config := newConfig(options...)
+	if len(config.Components) != 0 {
+		return nil, errors.New("NewLeaf: leaf timer cannot have components, otherwise use NewNode()")
+	}
+
+	component, ok := timer.components.lookup[name]
+	if !ok {
+		return nil, fmt.Errorf("NewLeaf: component %s not found", name)
+	}
+
+	return &baseTimer{
+		timeCache:       timer.timeCache,
+		config:          config,
+		parentComponent: component,
+		parent:          timer,
+	}, nil
+}
+
+func (timer *nodeTimer) MustLeaf(name string, options ...Option) Timer {
+	leaf, err := timer.NewLeaf(name, options...)
+	if err != nil {
+		panic(err)
+	}
+	return leaf
+}
+
+func (timer *nodeTimer) childStopped(duration time.Duration) {
+	if timer.parent == nil {
+		return
+	}
+
+	timer.parentComponent.spent.Add(int64(duration))
+	timer.parent.childStopped(duration)
+}
+
+func (timer *nodeTimer) Stats() map[string]time.Duration {
+	stats := make(map[string]time.Duration, len(timer.components.lookup))
+	for name, component := range timer.components.lookup {
+		stats[name] = time.Duration(component.spent.Load())
+	}
+
+	return stats
+}
+
+func (timer *nodeTimer) SumSpent() time.Duration {
```

Comment:
This, and `SumRemaining` as well as `SumExpired` are confusing me. Why are those not what's returned by `Spent`/`Remaining`/`Expired`? To me, as a user, I'd only care about these `Sum` variants as the other forms would give me confusing/misleading information?